src/engine/SCons/Tool/msvs.py

```diff
@@ -1285,7 +1317,7 @@ def printSources(self, hierarchy, kind, commonprefix, filter_name):
                 file = value
                 if commonprefix:
                     file = os.path.join(commonprefix, value)
-                file = os.path.normpath(file)
+                file = processSource(file)
```

Comment:
Why switch from normpath to abspath? If there's a good reason, I'd rather you collapse processSource's 1 line logic to be here rather than a function call which ends up obscuring the logic?

Reply:
When we print stuff out to a Visual Studio project file, we want everything to be absolute paths. Any relative path is dangerous because that is relative to the root directory, while the project file could be written to any random directory (e.g. build\devenv).

Reply:
I see so normpath is only good if written in the root directory? Could it use a relative path if written elsewhere? (relative to dir where project file written)?

Reply:
If the file is written in root dir, then it's safe to keep old logic?

Reply:
I reverted this part; I found that the consumer can provide absolute file paths if it needs it. The only place where I need to provide absolute paths is when generating include paths; the existing logic that just uses Dir nodes was causing include paths to incorrectly use pound-prefixed strings. I added comments around the parts that I changed and added some tests.